### PR TITLE
[Optimization] Read example pixels only when necessary 

### PR DESCRIPTION
**Pull request by Mr4k**
### Checklist

* [x] I have read the [Contributor Guide](../../CONTRIBUTING.md)
* [x] I have read and agree to the [Code of Conduct](../../CODE_OF_CONDUCT.md)
* [x] I have added a description of my changes and why I'd like them included in the section below

### Description of Changes

I'm not sure if this is the kind of contribution you are looking for but I did a little profiling (using instruments, screenshot below) and found out that (on my computer at least) a large amount of time was being taken up by the function `k_neighs_to_color_pattern` when creating the candidate patterns. It appears that looking up the pixels in the example images is somewhat costly and it is done in an inner loop which ends up contributing significantly to runtime.

instruments-profile 

To try to cut down on this cost I moved pixel lookups for the candidate's neighbors into the `better_match` function. I only read each pixel right before it needs to be used in the cost function. This means because are you already stopping a lot of the cost computations early (when the current candidate cost exceeds the smallest candidate cost so far) fewer pixel lookups are performed. This does not change the algorithm at all.

This results in around a 14% - 45% speed up (according to your benchmark test suite on my computer) depending on the example image(s) used and size of output texture. The average speed up seems to be more in the range of 14 - 25% (very unscientifically computed). I assume there are pathological cases where no performance gain could occur but I think they would be rare.

About my computer:
MacBook Pro (Retina, 13-inch, Early 2015)
Processor: 2.9 GHz Intel Core i5 (4 logical cores)
Memory: 8 GB 1867 MHz DDR3

Edit: Additionally tested with a Macbook Pro 2018
2.6 Ghz Intel Core i7 (12 logical cores)
32 GB 2400 MHz DDR4

I also tried to change the code minimally but there was some refactoring.

Disclaimer:
I have not tested this on a wide variety of devices or high end cpus

### Related Issues

I don't think this is related to any open issues.

### CODE REVIEW

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 1]
```diff
@@ -909,25 +891,14 @@ impl Generator {
...
-                                        k_neighs_to_color_pattern(
-                                            &cand.k_neighs,
-                                            image::Rgba([0, 0, 0, 255]),
-                                            &in_guides.example_guides,
-                                            &mut candidates_guide_patterns[cand_i],
-                                            false,
-                                        );
-
-                                        //get example pattern to compare to
-                                        k_neighs_to_color_pattern(
```
Comment (Mr4k) [inline lib/src/multires_stochastic_texture_synthesis.rs]: I'm not sure why this was in the candidates loop. It seemed to not matter so I moved it out

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 2]
```diff
@@ -1018,7 +987,49 @@ impl Generator {
...
+                .get_pixel(coord.x as u32, coord.y as u32)
+                .0[..4],
+        );
+    } else {
+        // if we have gone out of bounds, then just fill as outside color
+        neighbor_color.copy_from_slice(&outside_color.0[..]);
+    }
+}
+
+fn k_neighs_to_precomputed_reference_pattern(
```
Comment (Mr4k) [inline lib/src/multires_stochastic_texture_synthesis.rs]: renamed this to differentiate it from the now on the fly read candidate pixels, this function is now only called to build `my_pattern` and `my_guide_pattern`

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 3]
```diff
@@ -1117,45 +1106,63 @@ fn find_best_match<'a>(
...
-            if score >= current_best {
-                return None;
+            for (channel_n, &channel) in next_pixel.iter().enumerate() {
+                next_pixel_score +=
+                    guide_cost.get(my_precomputed_guide_pattern.0[i + channel_n], channel);
             }
         }
+        score += next_pixel_score * distance_gaussians[i];
+        if score >= current_best {
+            return None;
```
Comment (Mr4k) [inline lib/src/multires_stochastic_texture_synthesis.rs]: This is where the heart of the speed up is

File: lib/src/multires_stochastic_texture_synthesis.rs [hunk 4]
```diff
@@ -1117,45 +1106,63 @@ fn find_best_match<'a>(
...
+            );
 
-            if score >= current_best {
-                return None;
+            for (channel_n, &channel) in next_pixel.iter().enumerate() {
+                next_pixel_score +=
+                    guide_cost.get(my_precomputed_guide_pattern.0[i + channel_n], channel);
             }
         }
+        score += next_pixel_score * distance_gaussians[i];
```
Comment (Mr4k) [inline lib/src/multires_stochastic_texture_synthesis.rs]: should probably shrink distance gaussians from size (num neighbors) * 4 to just num neighbors as that is all I'm using here

**Comment by Jake-Shadle** [comment]
I unfortunately didn't have time to review this today, but I will check it out Monday, thanks for the PR!

**Comment by Mr4k** [comment]
Hey! First just wanted to say thanks so much for taking the time to review! 

Second of all I just wanted to make sure that I am not actually making your code worse :). One thing I notice when I look at the numbers in your benchmark is that it seems like performance has regressed on larger images which worries me a little. 

I'm curious what os / cpu you are using? I tried another macbook with an i7 and fewer cores (just 4 physical ones) and a digital ocean server with an Intel(R) Xeon(R) CPU E5-2697A v4 @ 2.60GHz and ended up seeing no performance regressions for larger images (still looks like a 25% speed up on average). I still have not found a non-intel processor to benchmark against.

I'm sure you're pretty busy but if you have the time I'd be really curious what results the following simple benchmarks give when run a couple times with and without this pr (I compared against the commit right before this one):
```
time cargo run --release -- --out out/01.jpg generate imgs/1.jpg

time cargo run --release -- --out out/01.jpg --out-size 1024 generate imgs/1.jpg

time cargo run --release -- --out out/01.jpg --out-size 2048 generate imgs/1.jpg

time cargo run --release -- --alpha 0.8 -o out/04.png transfer-style --style imgs/multiexample/4.jpg --guide imgs/tom.jpg
```

**Comment by Jake-Shadle** [comment]
I'm on an Intel Xeon 3.3Ghz on Linux. Ping @h3r2tic who has a Threadripper he can try this on.

**Comment by h3r2tic** [comment]
Ooooh, cool :) I'll test it when I'm back home!

**Comment by h3r2tic** [comment]
I ran it on my TR 2990WX:

```
BEFORE:

commit 3f30bea86b2bc6435cd56805d2ae2f4124d766ec (HEAD, tag: 0.7.1)
Author: Jake Shadle <jake.shadle@embark-studios.com>
Date:   Tue Nov 19 15:29:51 2019 +0100

    Release 0.7.1

time target/release/texture-synthesis.exe --out out/01.jpg generate imgs/1.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.485s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --out out/01.jpg generate imgs/1.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.472s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:00:15] ######################################## 100%
 stage   6 ######################################## 100%

real    0m15.673s
user    0m0.000s
sys     0m0.000s
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:01:04] ######################################## 100%
 stage   6 ######################################## 100%

real    1m5.580s
user    0m0.000s
sys     0m0.000s
time target/release/texture-synthesis.exe --alpha 0.8 -o out/04.png transfer-style --style imgs/multiexample/4.jpg --guide imgs/tom.jpg
[00:00:04] ######################################## 100%
 stage   6 ######################################## 100%

real    0m4.473s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --alpha 0.8 -o out/04.png transfer-style --style imgs/multiexample/4.jpg --guide imgs/tom.jpg
[00:00:04] ######################################## 100%
 stage   6 ######################################## 100%

real    0m4.459s
user    0m0.000s
sys     0m0.000s
```

```
AFTER:

commit 7196f75fc38963dadaa579c5a5bab3f5978eff9b (HEAD, origin/master, origin/HEAD, master)
Author: Jake Shadle <jake.shadle@embark-studios.com>
Date:   Mon Nov 25 14:23:21 2019 +0100

    Add CHANGELOG entry for PR#69

time target/release/texture-synthesis.exe --out out/01.jpg generate imgs/1.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.132s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --out out/01.jpg generate imgs/1.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.179s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:00:14] ######################################## 100%
 stage   6 ######################################## 100%

real    0m14.559s
user    0m0.000s
sys     0m0.015s
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:01:01] ######################################## 100%
 stage   6 ######################################## 100%

real    1m1.654s
user    0m0.000s
sys     0m0.000s
time target/release/texture-synthesis.exe --alpha 0.8 -o out/04.png transfer-style --style imgs/multiexample/4.jpg --guide imgs/tom.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.765s
user    0m0.000s
sys     0m0.000s
time target/release/texture-synthesis.exe --alpha 0.8 -o out/04.png transfer-style --style imgs/multiexample/4.jpg --guide imgs/tom.jpg
[00:00:03] ######################################## 100%
 stage   6 ######################################## 100%

real    0m3.788s
user    0m0.000s
sys     0m0.000s
```

**Comment by Jake-Shadle** [comment]
Interesting! Will have to look into why there's the slight regression on the larger ones on my xeon then.

**Comment by h3r2tic** [comment]
There's a slight regression here as well in the `--out out/01.jpg --out-size 2048 generate imgs/1.jpg` case.

**Comment by Mr4k** [comment]
Thanks for taking the time to benchmark this! 

If I'm reading those numbers correctly it looks like your results for the 2048 case are:
```
Before:
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:01:04] ######################################## 100%
 stage   6 ######################################## 100%

real    1m5.580s
user    0m0.000s
sys     0m0.000s

After:
time target/release/texture-synthesis.exe --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:01:01] ######################################## 100%
 stage   6 ######################################## 100%

real    1m1.654s
user    0m0.000s
sys     0m0.000s
```
So it looks like with this pr it ran in ~1:01 vs ~1:05 so it doesn't seem to be a regression. 

However those results are really minor (in fact I'd almost be worried they were noise) compared to what I've gotten on my test machines. 

For example here is my original test macbook:
```
Before:
time target/release/texture-synthesis --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:01:17] ######################################## 100%
 stage   6 ######################################## 100%

real	1m17.687s
user	3m47.141s
sys	0m1.396s
time target/release/texture-synthesis --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:01:17] ######################################## 100%
 stage   6 ######################################## 100%

real	1m17.759s
user	3m46.016s
sys	0m1.329s

time target/release/texture-synthesis --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:05:27] ######################################## 100%
 stage   6 ######################################## 100%

real	5m28.603s
user	15m35.845s
sys	0m5.621s
time target/release/texture-synthesis --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:05:23] ######################################## 100%
 stage   6 ######################################## 100%

real	5m25.009s
user	15m36.068s
sys	0m5.371s

After:
time target/release/texture-synthesis --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:01:01] ######################################## 100%
 stage   6 ######################################## 100%

real	1m1.418s
user	2m53.323s
sys	0m1.160s

time target/release/texture-synthesis --out out/01.jpg --out-size 1024 generate imgs/1.jpg
[00:01:01] ######################################## 100%
 stage   6 ######################################## 100%

real	1m1.640s
user	2m53.534s
sys	0m1.157s
time target/release/texture-synthesis --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:04:16] ######################################## 100%
 stage   6 ######################################## 100%

real	4m17.655s
user	12m2.829s
sys	0m4.699s
time target/release/texture-synthesis --out out/01.jpg --out-size 2048 generate imgs/1.jpg
[00:04:13] ######################################## 100%
 stage   6 ######################################## 100%

real	4m15.168s
user	11m58.227s
sys	0m4.631s
```
If you have the time I'd be curious to know (but I also don't want to take up too much of your time):
@h3r2tic what is the clock speed of your cpu? One thing I have not been able to bench this on is high clock speed cpus and while it's not a lot of data the mac above has a clock speed of 2.9ghz and I only see 20% gains as opposed to > 25% on the other computers I've run it on with slower cpus but more cores.
@Jake-Shadle what happens if you run these types of simple time commands as opposed to the bench suite?

**Comment by h3r2tic** [comment]
Ah, once again I find out that I can't even read xD You're obviously correct, @Mr4k, there was no regression there :)

It's a stock 2990wx (https://www.amd.com/en/products/cpu/amd-ryzen-threadripper-2990wx), so just 3GHz, but it does have 32 cores.

**Comment by Mr4k** [comment]
Did a few tests with 8 cores (16 threads) and a 2.8ghz intel xeon processor using a google cloud server and noticed that improvements were 16-18% instead of 20-30%. Should be able to test with even more cores soon. It seems conceivable that benefits drop off with large number of cores.

**Pull request merged**